neps/nep-0539.md

```diff
@@ -0,0 +1,741 @@
+      large receipts will be, we have to pessimistically reserve more space than
+      most receipts will actually need.
+    - Problem 2: If buffer space is shared globally, individual queues can still
+      grow really large, even indefinitely if we assume the number of shards
+      grows over time.
+    - Problem 3: If buffer space is on a per-shard basis, we run into deadlocks
+      when two shards have no more space left but both need to send to the other
+      shard to make progress.
+
+6. Require users to define which shards their transactions will touch and how
+   much gas is burnt in each. Then use this information for global scheduling
+   such that congestion is impossible.
+
+    - Problem 1: This requires lots of changes across the infrastructure stack.
+      It would take too long to implement as we are already facing congestion
+      problems today.
+    - Problem 2: This would have a strong impact on usability and it is unclear
+      if gas usage estimating services could close the gap to make it
+      acceptable.
+
+
+
+## Future possibilities
+
+While this proposal treats all requests the same, it sets the base for a proper
+transaction priority implementation. We co-designed this proposal with
+[NEP-541](https://github.com/near/NEPs/pull/541), which adds a transaction
+priority fee. On a very high level, the fee is used to auction off a part of the
+available gas per chunk to the highest bidders.
+
+We also expect that this proposal alone will not be the final solution for
+congestion control. Rather, we just want to build a solid foundation in this NEP
+and allow future optimization to take place on top of it.
+
+For example, estimations on how much gas is burnt on each shard could help with
+better load distribution in some cases.
+
+We also forsee that the round-robin scheduling of shard allowed to forward even
+under full congestion is not perfect. It is a key feature to make deadlocks
+provably impossible, since every shard is guaranteed to make a minimum progress
+after N rounds. But it could be beneficial to allocate more bandwidth to shards
+that actually have something to forward, or perhaps it would be better to stop
+forwarding anything for a while. The current proposal allows chunk producers to
+experiment with this without a protocol version change.
+
+Lastly, a future optimization could do better transaction rejection for meta
+transactions. Instead of looking only at the outer transaction receiver, we
+could also look at the receiver of the delegate action, which is most likely
+where most gas is going to be burnt, and use this for transaction rejection.
+
+## Consequences
+
+
+### 
+
+### Positive
+
+- Accepted transaction have lower latencies compared to today under congestion.
+- Storage and memory requirements on validator for storing receipts are bounded.
+
+### Neutral
+
+- More transactions are rejected at the chunk producer level.
```

Comment:
This is an important topic and a great point to be raised. Let me describe my thinking, which also reflects what we discussed in the last weekly congestion control meeting.

From a protocol point of view, we add a condition to reject transactions, meaning it would be invalid to add them to a chunk. What happens in the transaction pool isn't really relevant on that level, hence unspecified, or simply implementation choices.

The question regarding transaction pool queueing depends on at least two implementation choices.

- Are transactions removed from the transaction pool if they are rejected once due to congestion?
- Are RPC nodes filtering transaction according to current congestion before forwarding them to chunk producers?

Personally, I hope we can go for "Yes and Yes". RPC nodes should only accept a transaction if it has a high probability of being accepted. In case of rejection, users get immediate feedback. It's up to wallets to resend as often as appropriate. And chunk-producers should aggressively remove transaction that can't be included in chunks at the moment. For good UX, this rejection should also propagate back to the user through the RPC nodes. But if it's rare enough, we might get away without this in the initial deployment.

If we implement is as "No and No," then you are exactly right, we will generally queue more in the transaction pool until its size limit is reached.

"Yes and No" is also possible, which keeps the transaction pool short. But here it is even more important to propagate the rejection information back to the user, or else we have terrible UX since RPC nodes keep waiting for inclusion of the tx in a chunk when it was already dropped. But this back propagation is non-trivial to implement, given that multiple chunk producers are involved, so we might have to ship it later than the base congestion control mechanism.

"No and Yes" is also worth considering. Probably the easiest implementation without breaking UX. It's just odd because RPC nodes essentially control how large the transaction pool grows, rather than the chunk-producers that actually store it. But it could be a stepping stone towards "Yes and Yes".

Which ever way we go, I believe we don't need to change anything on the protocol level. And rejected transaction handling is fundamentally a client-side question that I believe should not be defined on the protocol level. In this discussion here, we only need to ensure the specification allows for decent client implementations.